File: Android/src/build-backups/kapt3-20260411-195127/stubs/debug/selfgemma/talk/data/roleplay/repository/RoomConversationRepository.java
package selfgemma.talk.data.roleplay.repository;

@javax.inject.Singleton()
@kotlin.Metadata(mv = {2, 2, 0}, k = 1, xi = 48, d1 = {"\u0000f\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\u0007\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u0002\n\u0002\b\f\n\u0002\u0010\b\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0004\n\u0002\u0018\u0002\n\u0002\b\u0006\b\u0007\u0018\u00002\u00020\u0001B)\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u0012\u0006\u0010\b\u001a\u00020\t\u00a2\u0006\u0004\b\n\u0010\u000bJ\u0014\u0010\f\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u000f0\u000e0\rH\u0016J\u001c\u0010\u0010\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00110\u000e0\r2\u0006\u0010\u0012\u001a\u00020\u0013H\u0016J\u001c\u0010\u0014\u001a\b\u0012\u0004\u0012\u00020\u00110\u000e2\u0006\u0010\u0012\u001a\u00020\u0013H\u0096@\u00a2\u0006\u0002\u0010\u0015J\u0018\u0010\u0016\u001a\u0004\u0018\u00010\u000f2\u0006\u0010\u0012\u001a\u00020\u0013H\u0096@\u00a2\u0006\u0002\u0010\u0015J(\u0010\u0017\u001a\u00020\u000f2\u0006\u0010\u0018\u001a\u00020\u00132\u0006\u0010\u0019\u001a\u00020\u00132\b\u0010\u001a\u001a\u0004\u0018\u00010\u001bH\u0096@\u00a2\u0006\u0002\u0010\u001cJ\u0016\u0010\u001d\u001a\u00020\u001e2\u0006\u0010\u001f\u001a\u00020\u000fH\u0096@\u00a2\u0006\u0002\u0010 J\u0016\u0010!\u001a\u00020\u001e2\u0006\u0010\u0012\u001a\u00020\u0013H\u0096@\u00a2\u0006\u0002\u0010\u0015J\u0016\u0010\"\u001a\u00020\u001e2\u0006\u0010\u0012\u001a\u00020\u0013H\u0096@\u00a2\u0006\u0002\u0010\u0015J\u0016\u0010#\u001a\u00020\u001e2\u0006\u0010$\u001a\u00020\u0011H\u0096@\u00a2\u0006\u0002\u0010%J\u0016\u0010&\u001a\u00020\u001e2\u0006\u0010$\u001a\u00020\u0011H\u0096@\u00a2\u0006\u0002\u0010%J$\u0010\'\u001a\u00020\u001e2\u0006\u0010\u0012\u001a\u00020\u00132\f\u0010(\u001a\b\u0012\u0004\u0012\u00020\u00110\u000eH\u0096@\u00a2\u0006\u0002\u0010)J\u0016\u0010*\u001a\u00020+2\u0006\u0010\u0012\u001a\u00020\u0013H\u0096@\u00a2\u0006\u0002\u0010\u0015J\u0018\u0010,\u001a\u0004\u0018\u00010-2\u0006\u0010\u0012\u001a\u00020\u0013H\u0096@\u00a2\u0006\u0002\u0010\u0015J\u0016\u0010.\u001a\u00020\u001e2\u0006\u0010/\u001a\u00020-H\u0096@\u00a2\u0006\u0002\u00100J\u001c\u00101\u001a\b\u0012\u0004\u0012\u0002020\u000e2\u0006\u0010\u0012\u001a\u00020\u0013H\u0096@\u00a2\u0006\u0002\u0010\u0015J\u0016\u00103\u001a\u00020\u001e2\u0006\u00104\u001a\u000202H\u0096@\u00a2\u0006\u0002\u00105J\u0016\u00106\u001a\u00020\u001e2\u0006\u0010$\u001a\u00020\u0011H\u0082@\u00a2\u0006\u0002\u0010%J\u0016\u00107\u001a\u00020\u001e2\u0006\u0010\u0012\u001a\u00020\u0013H\u0082@\u00a2\u0006\u0002\u0010\u0015R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u00068"}, d2 = {"Lselfgemma/talk/data/roleplay/repository/RoomConversationRepository;", "Lselfgemma/talk/domain/roleplay/repository/ConversationRepository;", "sessionDao", "Lselfgemma/talk/data/roleplay/db/dao/SessionDao;", "messageDao", "Lselfgemma/talk/data/roleplay/db/dao/MessageDao;", "sessionSummaryDao", "Lselfgemma/talk/data/roleplay/db/dao/SessionSummaryDao;", "sessionEventDao", "Lselfgemma/talk/data/roleplay/db/dao/SessionEventDao;", "<init>", "(Lselfgemma/talk/data/roleplay/db/dao/SessionDao;Lselfgemma/talk/data/roleplay/db/dao/MessageDao;Lselfgemma/talk/data/roleplay/db/dao/SessionSummaryDao;Lselfgemma/talk/data/roleplay/db/dao/SessionEventDao;)V", "observeSessions", "Lkotlinx/coroutines/flow/Flow;", "", "Lselfgemma/talk/domain/roleplay/model/Session;", "observeMessages", "Lselfgemma/talk/domain/roleplay/model/Message;", "sessionId", "", "listMessages", "(Ljava/lang/String;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "getSession", "createSession", "roleId", "modelId", "userProfile", "Lselfgemma/talk/domain/roleplay/model/StUserProfile;", "(Ljava/lang/String;Ljava/lang/String;Lselfgemma/talk/domain/roleplay/model/StUserProfile;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "updateSession", "", "session", "(Lselfgemma/talk/domain/roleplay/model/Session;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "archiveSession", "deleteSession", "appendMessage", "message", "(Lselfgemma/talk/domain/roleplay/model/Message;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "updateMessage", "replaceMessages", "messages", "(Ljava/lang/String;Ljava/util/List;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "nextMessageSeq", "", "getSummary", "Lselfgemma/talk/domain/roleplay/model/SessionSummary;", "upsertSummary", "summary", "(Lselfgemma/talk/domain/roleplay/model/SessionSummary;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "listEvents", "Lselfgemma/talk/domain/roleplay/model/SessionEvent;", "appendEvent", "event", "(Lselfgemma/talk/domain/roleplay/model/SessionEvent;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "syncSessionMetadata", "resyncSessionMetadata", "app_debug"})
public final class RoomConversationRepository implements selfgemma.talk.domain.roleplay.repository.ConversationRepository {
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.data.roleplay.db.dao.SessionDao sessionDao = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.data.roleplay.db.dao.MessageDao messageDao = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.data.roleplay.db.dao.SessionSummaryDao sessionSummaryDao = null;
    @org.jetbrains.annotations.NotNull()
    private final selfgemma.talk.data.roleplay.db.dao.SessionEventDao sessionEventDao = null;
    
    @javax.inject.Inject()
    public RoomConversationRepository(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.data.roleplay.db.dao.SessionDao sessionDao, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.data.roleplay.db.dao.MessageDao messageDao, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.data.roleplay.db.dao.SessionSummaryDao sessionSummaryDao, @org.jetbrains.annotations.NotNull()
    selfgemma.talk.data.roleplay.db.dao.SessionEventDao sessionEventDao) {
        super();
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.NotNull()
    public kotlinx.coroutines.flow.Flow<java.util.List<selfgemma.talk.domain.roleplay.model.Session>> observeSessions() {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.NotNull()
    public kotlinx.coroutines.flow.Flow<java.util.List<selfgemma.talk.domain.roleplay.model.Message>> observeMessages(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object listMessages(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super java.util.List<selfgemma.talk.domain.roleplay.model.Message>> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object getSession(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super selfgemma.talk.domain.roleplay.model.Session> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object createSession(@org.jetbrains.annotations.NotNull()
    java.lang.String roleId, @org.jetbrains.annotations.NotNull()
    java.lang.String modelId, @org.jetbrains.annotations.Nullable()
    selfgemma.talk.domain.roleplay.model.StUserProfile userProfile, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super selfgemma.talk.domain.roleplay.model.Session> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object updateSession(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.Session session, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object archiveSession(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object deleteSession(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object appendMessage(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.Message message, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object updateMessage(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.Message message, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object replaceMessages(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    java.util.List<selfgemma.talk.domain.roleplay.model.Message> messages, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object nextMessageSeq(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super java.lang.Integer> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object getSummary(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super selfgemma.talk.domain.roleplay.model.SessionSummary> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object upsertSummary(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.SessionSummary summary, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object listEvents(@org.jetbrains.annotations.NotNull()
    java.lang.String sessionId, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super java.util.List<selfgemma.talk.domain.roleplay.model.SessionEvent>> $completion) {
        return null;
    }
    
    @java.lang.Override()
    @org.jetbrains.annotations.Nullable()
    public java.lang.Object appendEvent(@org.jetbrains.annotations.NotNull()
    selfgemma.talk.domain.roleplay.model.SessionEvent event, @org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    private final java.lang.Object syncSessionMetadata(selfgemma.talk.domain.roleplay.model.Message message, kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    private final java.lang.Object resyncSessionMetadata(java.lang.String sessionId, kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
}

File: Android/src/app/src/main/java/selfgemma/talk/data/roleplay/repository/RoomConversationRepository.kt
package selfgemma.talk.data.roleplay.repository

import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import selfgemma.talk.data.roleplay.db.entity.SessionEntity
import selfgemma.talk.data.roleplay.db.dao.MessageDao
import selfgemma.talk.data.roleplay.db.dao.SessionDao
import selfgemma.talk.data.roleplay.db.dao.SessionEventDao
import selfgemma.talk.data.roleplay.db.dao.SessionSummaryDao
import selfgemma.talk.data.roleplay.mapper.toDomain
import selfgemma.talk.data.roleplay.mapper.toEntity
import selfgemma.talk.domain.roleplay.model.Message
import selfgemma.talk.domain.roleplay.model.MessageSide
import selfgemma.talk.domain.roleplay.model.MessageStatus
import selfgemma.talk.domain.roleplay.model.Session
import selfgemma.talk.domain.roleplay.model.SessionEvent
import selfgemma.talk.domain.roleplay.model.SessionSummary
import selfgemma.talk.domain.roleplay.model.StUserProfile
import selfgemma.talk.domain.roleplay.repository.ConversationRepository

private const val DEFAULT_SESSION_TITLE = "New Session"
private const val SESSION_TITLE_MAX_LENGTH = 48
private const val SESSION_EXCERPT_MAX_LENGTH = 140

@Singleton
class RoomConversationRepository
@Inject
constructor(
  private val sessionDao: SessionDao,
  private val messageDao: MessageDao,
  private val sessionSummaryDao: SessionSummaryDao,
  private val sessionEventDao: SessionEventDao,
) : ConversationRepository {

  override fun observeSessions(): Flow<List<Session>> {
    return sessionDao.observeActiveSessions().map { sessions -> sessions.map { it.toDomain() } }
  }

  override fun observeMessages(sessionId: String): Flow<List<Message>> {
    return messageDao.observeBySession(sessionId).map { messages -> messages.map { it.toDomain() } }
  }

  override suspend fun listMessages(sessionId: String): List<Message> {
    return messageDao.listBySession(sessionId).map { it.toDomain() }
  }

  override suspend fun getSession(sessionId: String): Session? {
    return sessionDao.getById(sessionId)?.toDomain()
  }

  override suspend fun createSession(roleId: String, modelId: String, userProfile: StUserProfile?): Session {
    val now = System.currentTimeMillis()
    val session =
      Session(
        id = UUID.randomUUID().toString(),
        roleId = roleId,
        title = DEFAULT_SESSION_TITLE,
        activeModelId = modelId,
        createdAt = now,
        updatedAt = now,
        lastMessageAt = now,
        sessionUserProfile = userProfile?.ensureDefaults(),
      )
    sessionDao.upsert(session.toEntity())
    return session
  }

  override suspend fun updateSession(session: Session) {
    sessionDao.upsert(session.toEntity())
  }

  override suspend fun archiveSession(sessionId: String) {
    sessionDao.archive(sessionId, System.currentTimeMillis())
  }

  override suspend fun deleteSession(sessionId: String) {
    sessionDao.delete(sessionId)
  }

  override suspend fun appendMessage(message: Message) {
    messageDao.insert(message.toEntity())
    syncSessionMetadata(message)
  }

  override suspend fun updateMessage(message: Message) {
    messageDao.update(message.toEntity())
    syncSessionMetadata(message)
  }

  override suspend fun replaceMessages(sessionId: String, messages: List<Message>) {
    messageDao.deleteBySession(sessionId)
    if (messages.isNotEmpty()) {
      messageDao.insertAll(messages.map { it.toEntity() })
    }
    resyncSessionMetadata(sessionId)
  }

  override suspend fun nextMessageSeq(sessionId: String): Int {
    return messageDao.getMaxSeq(sessionId) + 1
  }

  override suspend fun getSummary(sessionId: String): SessionSummary? {
    return sessionSummaryDao.getBySessionId(sessionId)?.toDomain()
  }

  override suspend fun upsertSummary(summary: SessionSummary) {
    sessionSummaryDao.upsert(summary.toEntity())

    val session = sessionDao.getById(summary.sessionId) ?: return
    sessionDao.upsert(
      session.copy(
        lastSummary = summary.summaryText,
        summaryVersion = summary.version,
        updatedAt = summary.updatedAt,
      )
    )
  }

  override suspend fun listEvents(sessionId: String): List<SessionEvent> {
    return sessionEventDao.listBySession(sessionId).map { it.toDomain() }
  }

  override suspend fun appendEvent(event: SessionEvent) {
    sessionEventDao.insert(event.toEntity())
  }

  private suspend fun syncSessionMetadata(message: Message) {
    val session = sessionDao.getById(message.sessionId) ?: return
    val completedTurns =
      if (message.side == MessageSide.ASSISTANT && message.status == MessageStatus.COMPLETED) {
        messageDao.countBySideAndStatus(
          sessionId = message.sessionId,
          side = MessageSide.ASSISTANT,
          status = MessageStatus.COMPLETED,
        )
      } else {
        session.turnCount
      }

    sessionDao.upsert(session.mergeWithMessage(message, completedTurns))
  }

  private suspend fun resyncSessionMetadata(sessionId: String) {
    val session = sessionDao.getById(sessionId) ?: return
    val messages = messageDao.listBySession(sessionId).map { it.toDomain() }
    val updatedSession =
      if (messages.isEmpty()) {
        session.copy(
          title = DEFAULT_SESSION_TITLE,
          updatedAt = maxOf(session.updatedAt, session.createdAt),
          lastMessageAt = session.createdAt,
          lastUserMessageExcerpt = null,
          lastAssistantMessageExcerpt = null,
          turnCount = 0,
        )
      } else {
        messages.fold(
          initial = session.copy(
            title = DEFAULT_SESSION_TITLE,
            lastUserMessageExcerpt = null,
            lastAssistantMessageExcerpt = null,
            turnCount = 0,
          )
        ) { accumulator, message ->
          accumulator.mergeWithMessage(
            message = message,
            completedTurns =
              if (message.side == MessageSide.ASSISTANT && message.status == MessageStatus.COMPLETED) {
                accumulator.turnCount + 1
              } else {
                accumulator.turnCount
              },
          )
        }
      }
    sessionDao.upsert(updatedSession)
  }
}

private fun SessionEntity.mergeWithMessage(message: Message, completedTurns: Int): SessionEntity {
  return copy(
    title = deriveSessionTitle(title, message),
    updatedAt = maxOf(updatedAt, message.updatedAt),
    lastMessageAt = maxOf(lastMessageAt, message.updatedAt),
    lastUserMessageExcerpt =
      if (message.side == MessageSide.USER) {
        message.content.toExcerpt()
      } else {
        lastUserMessageExcerpt
      },
    lastAssistantMessageExcerpt =
      if (message.side == MessageSide.ASSISTANT && message.content.isNotBlank()) {
        message.content.toExcerpt()
      } else {
        lastAssistantMessageExcerpt
      },
    turnCount = completedTurns,
  )
}

private fun deriveSessionTitle(currentTitle: String, message: Message): String {
  if (currentTitle != DEFAULT_SESSION_TITLE) {
    return currentTitle
  }

  if (message.side != MessageSide.USER || message.content.isBlank()) {
    return currentTitle
  }

  return message.content.trim().replace("\n", " ").take(SESSION_TITLE_MAX_LENGTH)
}

private fun String.toExcerpt(): String {
  return trim().replace("\n", " ").take(SESSION_EXCERPT_MAX_LENGTH).ifBlank { this }
}
